src/passes/OptimizeInstructions.cpp

```diff
@@ -1334,104 +1334,135 @@ struct OptimizeInstructions
+      // though, that we cannot do this if we cannot replace the current type
+      // with the reference's type.) We can also do this if this is a static
+      // cast: in that case, all we need to know about are the types.
+      if (HeapType::isSubType(curr->ref->type.getHeapType(), intendedType)) {
+        if (curr->rtt) {
           replaceCurrent(getResultOfFirst(curr->ref,
                                           builder.makeDrop(curr->rtt),
                                           getFunction(),
                                           getModule(),
                                           passOptions));
-          return;
+        } else {
+          replaceCurrent(curr->ref);
         }
+        return;
       }
+    }
 
-      // Repeated identical ref.cast operations are unnecessary, if using the
-      // exact same rtt - the result will be the same. Find the immediate child
-      // cast, if there is one, and see if it is identical.
-      // TODO: Look even further through incompatible casts?
-      auto* ref = curr->ref;
-      while (!ref->is<RefCast>()) {
-        auto* last = ref;
-        // RefCast falls through the value, so instead of calling
-        // getFallthrough() to look through all fallthroughs, we must iterate
-        // manually. Keep going until we reach either the end of things
-        // falling-through, or a cast.
-        ref =
-          Properties::getImmediateFallthrough(ref, passOptions, *getModule());
-        if (ref == last) {
-          break;
-        }
+    // Repeated identical ref.cast operations are unnecessary. First, find the
+    // immediate child cast, if there is one.
+    // TODO: Look even further through incompatible casts?
+    auto* ref = curr->ref;
+    while (!ref->is<RefCast>()) {
+      auto* last = ref;
+      // RefCast falls through the value, so instead of calling
+      // getFallthrough() to look through all fallthroughs, we must iterate
+      // manually. Keep going until we reach either the end of things
+      // falling-through, or a cast.
+      ref = Properties::getImmediateFallthrough(ref, passOptions, *getModule());
+      if (ref == last) {
+        break;
       }
-      if (auto* child = ref->dynCast<RefCast>()) {
+    }
+    if (auto* child = ref->dynCast<RefCast>()) {
+      if (curr->rtt && child->rtt) {
         // Check if the casts are identical.
         if (ExpressionAnalyzer::equal(curr->rtt, child->rtt) &&
             !EffectAnalyzer(passOptions, *getModule(), curr->rtt)
                .hasSideEffects()) {
           replaceCurrent(curr->ref);
           return;
         }
+      } else if (!curr->rtt && !child->rtt) {
+        // Repeated static casts can be removed, leaving just the most demanding
+        // of them.
+        auto childIntendedType = child->getIntendedType();
+        if (HeapType::isSubType(intendedType, childIntendedType)) {
+          // Skip the child.
+          curr->ref = child->ref;
```

Comment:
`replaceCurrent` replaces the current node (which is not needed here), and also in this file it will further optimize the new output as if we just reached it. But in this case we don't need to do any more work, since the parent is the stricter: if the child had any stricter child, it would have been replace by it already during the postorder walk. (And the new tests verify that now as mentioned above.)